lib/src/podman_ostree.rs

```diff
@@ -0,0 +1,296 @@
+//! # Mapping between podman/containers-storage: and ostree
+//!
+//! The common container storage model is to store blobs (layers) as unpacked directories,
+//! and use the Linux `overlayfs` to merge them dynamically.
+//!
+//! However, today the `ostree-prepare-root` model as used by ostree expects a final flattened
+//! filesystem tree; and crucially we need to perform SELinux labeling.  At the moment, because
+//! ostree again works on just a plain directory, we need to "physically" change the on-disk
+//! xattrs of the target files.
+//!
+//! That said, there is work in ostree to use composefs, which will add a huge amount of flexibility;
+//! we can generate an erofs blob dynamically with the target labels.
+//!
+//! Even more than that however the ostree core currently expects an ostree commit object to be backing
+//! the filesystem tree; this is how it handles garbage collection, inspects metadata, etc.  Parts
+//! of bootc rely on this too today.
+//!
+//! ## Disadvantages
+//!
+//! One notable disadvantage of this model is that we're storing file *references* twice,
+//! which means the ostree deduplication is pointless.  In theory this is fixable by going back
+//! and changing the containers-storage files, but...
+//!
+//! ## Medium term: Unify containers-storage and ostree with composefs
```

Comment:
One thought here...we could bite the bullet and add Go code to this project, and vendor c/storage and c/image.  It'd give us the ability to even handle patches for those if need be that we release before waiting for podman or skopeo.

It would also help clear the way for us to do things like copy the podman code for quadlet files to do stuff like https://github.com/containers/podman/issues/22785 right away, again without depending on podman changes in the short term.

Notable tradeoffs either way.  But, the more I think about it the more I lean that direction.  Does anyone have strong feelings about this?

Reply:
I'd rather stick with the tools. The tools are always released in sync to make sure that no change to c/storage, for instance, causes any unexpected regression. OCP ran into issues because of using different versions of c/storage in CRI-O and Podman.